### PR TITLE
[LETS-212] Add passive transaction server configuration

### PR DESCRIPTION
**Pull request by adrianbrab**
http://jira.cubrid.org/browse/LETS-212

Transaction server needs to be configured to be an active transaction server, or a passive transaction server (replica/standby). Add transaction_server_type parameter to configure it.

- Added enum class transaction_server_type with the values ACTIVE, PASSIVE and the transaction_server_type_to_string function.
- Added enum class transaction_server_type_config with the values ACTIVE, REPLICA and STANDBY and the transaction_server_type_config_to_string function
- Added transaction_server_type system parameter, with default value active
- Added g_transaction_server_type
- Added the functions is_active_transaction_server and is_passive_transaction_server

Transaction server type is visible in the server error log with the following message:
```
Time: 10/06/21 18:20:27.601 - DEBUG *** file /home/adrian/cubridScalabil/cubrid/src/base/server_type.cpp, line 144
Transaction server type: passive
```

### CODE REVIEW

File: src/base/server_type.cpp [hunk 1]
```diff
@@ -146,6 +179,17 @@ void finalize_server_type ()
     }
 }
 
+bool is_active_transaction_server ()
+
```
Comment (ribram) [inline src/base/server_type.cpp:188]:
```suggestion
```

File: src/base/server_type.cpp [hunk 2]
```diff
@@ -146,6 +179,17 @@ void finalize_server_type ()
...
+{
+  return g_transaction_server_type == transaction_server_type::PASSIVE;
```
Comment (ribram) [inline src/base/server_type.cpp]:
```suggestion
  return g_server_type == SERVER_TYPE_TRANSACTION && g_transaction_server_type == transaction_server_type::PASSIVE;
```

File: src/base/server_type.cpp [hunk 3]
```diff
@@ -146,6 +179,17 @@ void finalize_server_type ()
...
+}
+
+bool is_passive_transaction_server ()
+{
+  return g_transaction_server_type == transaction_server_type::PASSIVE;
```
Comment (ribram) [inline src/base/server_type.cpp]:
```suggestion
  return g_server_type == SERVER_TYPE_TRANSACTION && g_transaction_server_type == transaction_server_type::PASSIVE;
```

File: src/base/server_type.cpp [hunk 4]
```diff
@@ -146,6 +179,17 @@ void finalize_server_type ()
...
+}
+
```
Comment (ribram) [inline src/base/server_type.cpp:208]: Please also add the functions:

`is_transaction_server ()` = `g_server_type == SERVER_TYPE_TRANSACTION`
`is_page_server ()` = `g_server_type == SERVER_TYPE_PAGE`
`is_passive_server ()` = `is_page_server ()` || `is_passive_transaction_server ()`

File: src/base/server_type.cpp [hunk 5]
```diff
@@ -146,6 +179,31 @@ void finalize_server_type ()
...
+bool is_page_server ()
+{
+  return g_server_type == SERVER_TYPE_PAGE;
+}
+
+bool is_passive_transaction_server ()
+{
+  return is_transaction_server() && g_transaction_server_type == transaction_server_type::PASSIVE;
```
Comment (ribram) [inline src/base/server_type.cpp]:
```suggestion
  return is_transaction_server () && g_transaction_server_type == transaction_server_type::PASSIVE;
```

File: src/base/system_parameter.c [hunk 6]
```diff
@@ -2378,6 +2379,12 @@ static int prm_server_type_default = (int) server_type_config::TRANSACTION;
 static int prm_server_type_lower = (int) server_type_config::TRANSACTION;
 static int prm_server_type_upper = (int) server_type_config::SINGLE_NODE;
 static unsigned int prm_server_type_flag = 0;
+
+int PRM_TRANSACTION_SERVER_TYPE = (int) transaction_server_type_config::ACTIVE;
+static int prm_transaction_server_type_default = (int) transaction_server_type_config::ACTIVE;
```
Comment (cristiarg) [inline src/base/system_parameter.c]:
```suggestion
static int prm_transaction_server_type_default = (int) transaction_server_type_config::ACTIVE;
int PRM_TRANSACTION_SERVER_TYPE = prm_transaction_server_type_default;
```

File: src/base/server_type.cpp [hunk 7]
```diff
@@ -111,6 +138,12 @@ int init_server_type (const char *db_name)
   if (er_code == NO_ERROR)
     {
       er_log_debug (ARG_FILE_LINE, "Starting server type: %s\n", server_type_to_string (get_server_type ()));
+
+      if (g_server_type == SERVER_TYPE_TRANSACTION)
```
Comment (cristiarg) [inline src/base/server_type.cpp:140]: Would be better to only have one log line that describes everything such that a single grep reveals the whole status/type.

File: src/base/server_type.hpp [hunk 8]
```diff
@@ -24,6 +24,12 @@
 int init_server_type (const char *db_name);
 void finalize_server_type ();
 SERVER_TYPE get_server_type ();
+transaction_server_type get_transaction_server_type ();
+bool is_active_transaction_server();
```
Comment (cristiarg) [inline src/base/server_type.hpp]: Order these functions - at least in the header - by their "composability" factor from simple to complex.

File: src/base/system_parameter.c [hunk 9]
```diff
@@ -6488,6 +6507,12 @@ static KEYVAL server_type_words[] = {
   {server_type_config_to_string (server_type_config::PAGE), (int) server_type_config::PAGE},
   {server_type_config_to_string (server_type_config::SINGLE_NODE), (int) server_type_config::SINGLE_NODE}
 };
+
+static KEYVAL transaction_server_type_words[] = {
```
Comment (cristiarg) [inline src/base/system_parameter.c:6511]:
```suggestion
static constexpr KEYVAL transaction_server_type_words[] = {
```
All mappings in this file can be `constexpr` - but maybe is beyond the scope of this PR:
```
static constexpr KEYVAL boolean_words[] = {
static constexpr KEYVAL er_log_level_words[] = {
static constexpr KEYVAL isolation_level_words[] = {
static constexpr KEYVAL pgbuf_debug_page_validation_level_words[] = {
static constexpr KEYVAL null_words[] = {
static constexpr KEYVAL ha_mode_words[] = {
static constexpr KEYVAL ha_server_state_words[] = {
static constexpr KEYVAL ha_log_applier_state_words[] = {
static constexpr KEYVAL compat_words[] = {
static constexpr KEYVAL check_peer_alive_words[] = {
static constexpr KEYVAL query_trace_format_words[] = {
static constexpr KEYVAL fi_test_words[] = {
static constexpr KEYVAL ha_repl_filter_type_words[] = {
static constexpr KEYVAL tde_algorithm_words[] = {
static constexpr KEYVAL server_type_words[] = {
```

File: src/base/system_parameter.c [hunk 10]
```diff
@@ -8440,6 +8465,12 @@ prm_print (const SYSPRM_PARAM * prm, char *buf, size_t len, PRM_PRINT_MODE print
 	{
 	  keyvalp = prm_keyword (PRM_GET_INT (prm->value), NULL, server_type_words, DIM (server_type_words));
 	}
+      else if (intl_mbs_casecmp (prm->name, PRM_NAME_TRANSACTION_SERVER_TYPE) == 0)
+	{
+	  keyvalp =
```
Comment (cristiarg) [inline src/base/system_parameter.c]: alignment

File: src/base/system_parameter.c (hunk 6 above)
Comment (adrianbrab) [reply src/base/system_parameter.c]: Changed, and also made the same change for server_type for consistency.

File: src/base/server_type.hpp (hunk 8 above)
Comment (adrianbrab) [reply src/base/server_type.hpp]: Ordered now.

File: src/base/system_parameter.c (hunk 9 above)
Comment (adrianbrab) [reply src/base/system_parameter.c:6511]: i think a small issue is best suited for this change.

**Review by cristiarg — APPROVED** [review]

File: src/base/server_type.cpp [hunk 11]
```diff
@@ -146,6 +181,31 @@ void finalize_server_type ()
     }
 }
 
+bool is_active_transaction_server ()
+{
+  return is_transaction_server() && g_transaction_server_type == transaction_server_type::ACTIVE;
```
Comment (ribram) [inline src/base/server_type.cpp]:
```suggestion
  return is_transaction_server () && g_transaction_server_type == transaction_server_type::ACTIVE;
```

File: src/base/server_type.hpp (hunk 8 above)
Comment (ribram) [reply src/base/server_type.hpp]:
```suggestion
bool is_active_transaction_server ();
```
```suggestion
bool is_active_transaction_server();
```

File: src/base/system_parameter.c (hunk 9 above)
Comment (ribram) [reply src/base/system_parameter.c:6511]: Please limit the review suggestions in the purpose of the patch, or to things that may have a meaningful impact. The legacy code is still 99% C and there are thousands of examples like this, that may be modernized and aligned to new C++ standards.

**Review by ribram — APPROVED** [review]

**Pull request merged**